libcudacxx/include/cuda/__cmath/fast_modulo_division.h

```diff
@@ -0,0 +1,181 @@
+      {
+        __multiplier = 0;
+        return;
+      }
+      auto __k              = __num_bits_v<_Tp> + __shift;
+      __multiplier          = ((__larger_t{1} << __k) + (__larger_t{1} << __shift)) / __divisor;
+      auto __multiplier_low = (__larger_t{1} << __k) / __divisor;
+      __add                 = (__multiplier_low == __multiplier);
+    }
+  }
+
+  template <typename _Up>
+  [[nodiscard]] _CCCL_API friend _CUDA_VSTD::common_type_t<_Tp, _Up>
+  operator/(_Up __dividend, fast_mod_div<_Tp> __divisor1) noexcept
+  {
+    using namespace _CUDA_VSTD;
+    static_assert(__cccl_is_integer_v<_Up> && sizeof(_Up) <= 8, "unsupported type");
+    static_assert(sizeof(_Up) <= sizeof(_Tp), "dividend type must be not larger than the divisor type");
+    static_assert(!(is_unsigned_v<_Up> && is_signed_v<_Tp>), "unsupported types");
+    if constexpr (is_signed_v<_Up>)
+    {
+      _CCCL_ASSERT(__dividend >= 0, "dividend must be non-negative");
+    }
+    using __common_t    = common_type_t<_Tp, _Up>;
+    const auto __div    = __divisor1.__divisor; // cannot use structure binding because of clang-14
+    const auto __mul    = __divisor1.__multiplier;
+    const auto __add_   = __divisor1.__add; // cannot use __add because of shadowing warning with clang-cuda
+    const auto __shift_ = __divisor1.__shift;
+    if (!_IsDivisorNotOne && __div == 1)
+    {
+      return __dividend;
+    }
+    if constexpr (is_unsigned_v<_Tp>)
+    {
+      if (__mul == 0) // divisor is a power of two
+      {
+        return static_cast<__common_t>(__dividend >> __shift_);
+      }
+      if (__dividend != numeric_limits<_Up>::max()) // avoid overflow
+      {
+        __dividend += __add_;
+      }
+    }
+    auto __higher_bits = ::cuda::__multiply_extract_higher_bits(__dividend, __mul);
+    auto __quotient    = __higher_bits >> __shift_;
+    _CCCL_ASSERT(__quotient == (static_cast<__unsigned_t>(__dividend) - __add_) / __div, "wrong __quotient");
+    return static_cast<__common_t>(__quotient);
+  }
+
+  template <typename _Up>
+  [[nodiscard]] _CCCL_API friend _CUDA_VSTD::common_type_t<_Tp, _Up>
+  operator%(_Up __dividend, fast_mod_div<_Tp> __divisor1) noexcept
+  {
+    return __dividend - (__dividend / __divisor1) * __divisor1.__divisor;
+  }
+
+  [[nodiscard]] _CCCL_API operator _Tp() const noexcept
+  {
+    return static_cast<_Tp>(__divisor);
+  }
+
+private:
+  _Tp __divisor             = 1;
+  __unsigned_t __multiplier = 0;
+  int __add                 = 0;
+  int __shift               = 0;
+};
+
+/***********************************************************************************************************************
+ * Non-member functions
+ **********************************************************************************************************************/
+
+template <typename _Tp, typename _Up>
+[[nodiscard]] _CCCL_API _CUDA_VSTD::pair<_Tp, _Up> div(_Tp __dividend, fast_mod_div<_Up> __divisor) noexcept
```

Comment:
I rather had this as `_Up` and then convert to `fast_mod_div` internally. 

Currently the user would have to do that by hand